modin/dataframe/groupby.py

```diff
@@ -81,48 +62,96 @@ def __getattr__(self, key):
                     "github.com/ray-project/ray.")
             raise e
 
+    _index_grouped_cache = None
+
+    @property
+    def _index_grouped(self):
+        if self._index_grouped_cache is None:
+            if self._axis == 0:
+                self._index_grouped_cache = pandas.Series(
+                    np.zeros(len(self._index), dtype=np.uint8),
+                    index=self._index).groupby(by=self._by, sort=self._sort)
+            else:
+                self._index_grouped_cache = pandas.Series(
+                    np.zeros(len(self._columns), dtype=np.uint8),
+                    index=self._columns).groupby(by=self._by, sort=self._sort)
+
+        return self._index_grouped_cache
+
+    _keys_and_values_cache = None
+
+    @property
+    def _keys_and_values(self):
+        if self._keys_and_values_cache is None:
+            self._keys_and_values_cache = \
+                [(k, v) for k, v in self._index_grouped]
+        return self._keys_and_values_cache
+
+    @property
+    def _grouped_partitions(self):
+
+        # It is expensive to put this multiple times, so let's just put it once
+        remote_by = ray.put(self._by)
+
+        if len(self) > 1:
```

Comment:
Can you change `len(self)` to `len(self._index_grouped)`, the same effect but more readable.